src/provider/provider_ctl_stats_impl.h

```diff
@@ -0,0 +1,128 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ctl/ctl.h"
+#include <utils/utils_assert.h>
+
+static int CTL_READ_HANDLER(peak_memory)(void *ctx,
+                                         umf_ctl_query_source_t source,
+                                         void *arg,
+                                         umf_ctl_index_utlist_t *indexes,
+                                         const char *extra_name,
+                                         umf_ctl_query_type_t query_type) {
+    /* suppress unused-parameter errors */
+    (void)source, (void)indexes, (void)ctx, (void)extra_name, (void)query_type;
+
+    size_t *arg_out = arg;
+    CTL_PROVIDER_TYPE *provider = (CTL_PROVIDER_TYPE *)ctx;
+    COMPILE_ERROR_ON(sizeof(provider->stats.peak_memory) != sizeof(uint64_t));
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.peak_memory,
+                                  (uint64_t *)arg_out);
+    return 0;
+}
+
+static int CTL_READ_HANDLER(allocated_memory)(void *ctx,
+                                              umf_ctl_query_source_t source,
+                                              void *arg,
+                                              umf_ctl_index_utlist_t *indexes,
+                                              const char *extra_name,
+                                              umf_ctl_query_type_t query_type) {
+    /* suppress unused-parameter errors */
+    (void)source, (void)indexes, (void)ctx, (void)extra_name, (void)query_type;
+
+    size_t *arg_out = arg;
+    CTL_PROVIDER_TYPE *provider = (CTL_PROVIDER_TYPE *)ctx;
+    COMPILE_ERROR_ON(sizeof(provider->stats.allocated_memory) !=
+                     sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(*arg_out) != sizeof(uint64_t));
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.allocated_memory,
+                                  (uint64_t *)arg_out);
+    return 0;
+}
+
+static int CTL_RUNNABLE_HANDLER(reset)(void *ctx, umf_ctl_query_source_t source,
+                                       void *arg,
+                                       umf_ctl_index_utlist_t *indexes,
+                                       const char *extra_name,
+                                       umf_ctl_query_type_t query_type) {
+    /* suppress unused-parameter errors */
+    (void)source, (void)indexes, (void)arg, (void)extra_name, (void)query_type;
+
+    CTL_PROVIDER_TYPE *provider = (CTL_PROVIDER_TYPE *)ctx;
+    size_t allocated;
+
+    COMPILE_ERROR_ON(sizeof(provider->stats.allocated_memory) !=
+                     sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(allocated) != sizeof(uint64_t));
+
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.allocated_memory,
+                                  (uint64_t *)&allocated);
+    utils_atomic_store_release_u64((uint64_t *)&provider->stats.peak_memory,
+                                   (uint64_t)allocated);
+
+    return 0;
+}
+
+static const umf_ctl_node_t CTL_NODE(stats)[] = {
```

Comment:
could you move this to the top of the file? what would be the type of these vars?

Reply:
nope - the order is important, do to symbol visibility limitations.(you have to declare symbols before use)